### PR TITLE
Use PostRepository.search to fetch posts and pages for dashboard

### LINKED ISSUE #21812: Dashboard loads all pages on a site instead of just three
### Background

Dashboard has a "Pages" card, which displays the current user's most recent three pages. The relevant code should load three pages because that's all this feature requires. And that's [how the code was written](https://github.com/wordpress-mobile/WordPress-iOS/blob/23.4/WordPress/Classes/ViewRelated/Blog/Blog%20Dashboard/Helpers/DashboardPostsSyncManager.swift#L108). However, the app actually loads all pages, instead of just three pages.

### Expected behavior

The API requests initiated by the dashboard "Pages" card fetch three pages.

### Actual behavior

The API requests initiated by the dashboard "Pages" card fetch all pages that are created by the user.

### Steps to reproduce the behavior

Prepare a site with 10+ posts (to make the issue more obvious, the more the better). Make sure all posts are authored by the test user.

1. Add a log at [the fetching posts/pages code](https://github.com/wordpress-mobile/WordPress-iOS/blob/23.4/WordPress/Classes/Services/PostService.m#L122). Or better yet, set up a proxy server (mitmproxy or Charles) to see all HTTP traffic sent from the app.
2. Launch the app and go to the home screen on My Sites tab.
3. Make sure "Pages" card is visible.
4. Pull to refresh, which should make the app sends API request to fetch pages.
5. Depending on your setup in step 1, check out the console logs in Xcode or `GET /site/%s/posts` requests captured by your proxy server.

##### Tested on [device], iOS [version], Jetpack iOS / WordPress iOS [version]
Jetpack 23.4

### PR DESCRIPTION
**Pull request by crazytonyli**
Fixes #21812.

The root cause of the issue is `PostService.syncPosts` function makes API requests fetch _all pages_ that match the given criteria, regardless what the `number` parameter is. The new `PostRepository.search` (added in https://github.com/wordpress-mobile/WordPress-iOS/pull/21814) was designed to fix this issue.

To test, verify the issue is indeed fixed.

Large part of this PR's changes are in the unit tests, because I have switched to use an existing HTTP stubs to get posts list, instead of creating a fake `PostRepository`.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Verifying the issue is fixed.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist: N/A

### CODE REVIEW

File: WordPress/WordPressTest/Dashboard/DashboardPostsSyncManagerTests.swift [hunk 1]
```diff
@@ -26,101 +36,95 @@ class DashboardPostsSyncManagerTests: CoreDataTestCase {
...
+        stubGetPostsList(type: "post", total: 50)
 
-        let manager = DashboardPostsSyncManager(postService: postService, blogService: blogService)
+        let manager = DashboardPostsSyncManager(postRepository: postRepository, blogService: blogService)
         let listener = SyncManagerListenerMock()
         manager.addListener(listener)
 
         // When
         manager.syncPosts(blog: blog, postType: .post, statuses: draftStatuses)
+        wait(for: [expectation(that: \.postsSyncedCalled, on: listener, willEqual: true)], timeout: 0.1)
```
Comment (mokagio) [inline WordPress/WordPressTest/Dashboard/DashboardPostsSyncManagerTests.swift:47]: First time I see this style. Neat.

File: WordPress/WordPressTest/Dashboard/DashboardPostsSyncManagerTests.swift [hunk 2]
```diff
@@ -179,14 +183,13 @@ class DashboardPostsSyncManagerTests: CoreDataTestCase {
...
-        XCTAssertFalse(postService.syncPostsCalled)
     }
 
 }
 
-class SyncManagerListenerMock: DashboardPostsSyncManagerListener {
+class SyncManagerListenerMock: NSObject, DashboardPostsSyncManagerListener {
 
-    var postsSyncedCalled = false
+    @objc dynamic var postsSyncedCalled = false
```
Comment (mokagio) [inline WordPress/WordPressTest/Dashboard/DashboardPostsSyncManagerTests.swift]: What do you think of having these properties `private(set)`?

I like if from a safety point of view, just to make sure that a test doesn't accidentally update them directly, as opposed to the side effect of the SUT interacting with the test double.

In practice, I acknowledge it's much of a muchness and requires extra typing (unless you get AI to help, but that's hit and miss).

---

This is not a suggestion for the diff. More of a "should we have a convention" question.

File: WordPress/WordPressTest/PostRepositoryPostsListTests.swift [hunk 3]
```diff
@@ -63,7 +63,7 @@ class PostRepositoryPostsListTests: CoreDataTestCase {
     }
 
     func testFetchingAllPages() async throws {
-        try await preparePostsList(type: "page", total: 1_120)
+        stubGetPostsList(type: "page", total: 1_120)
```
Comment (mokagio) [inline WordPress/WordPressTest/PostRepositoryPostsListTests.swift]: This test takes ~7s to run in CI, according to [Test Analytics](https://buildkite.com/organizations/automattic/analytics/suites/wordpress-ios/runs/325ce80a-8caf-4e29-aba3-612b72517cb9). The 1,120 amount was there already, but I wonder if the value we get from it is worth the speed bump.

image

**Review by mokagio — APPROVED** [review]
>  I have switched to use an existing HTTP stubs to get posts list, instead of creating a fake `PostRepository`.

What was your rationale for this change? I'm in particular interested in using HTTP stubs over having a test double for `PostRepository`.

---

I run out of time before a mid-day stop to test manually, but the unit tests here do a great job at ensuring correctness.

Approving to unblock, but I think it'd be best to address that 7s test before merging. In the current setup, 7s is not a bottleneck. But that's no reason to avoid improving upon it if we can.

File: WordPress/WordPressTest/Dashboard/DashboardPostsSyncManagerTests.swift (hunk 2 above)
Comment (crazytonyli) [reply WordPress/WordPressTest/Dashboard/DashboardPostsSyncManagerTests.swift]: Sounds reasonable. These properties are meant to be private(set) after all. I doubt they were intentionally declared as allowing other to change their value.

**Comment by crazytonyli** [comment]
> What was your rationale for this change? I'm in particular interested in using HTTP stubs over having a test double for PostRepository.

I believe you meat having a test double for `PostServiceRemote`, which is the one that makes API requests?

HTTP stubs are easier to share. As you can see in this PR, I only moved the original `preparePostsList` implementation, without any real code changes. And, now it can serve both test cases without the test cases making any accommodating changes. Also, in this particular case, we just need a certain amount of posts and that's handled in one place - the `stubGetPostsList` function.

We can certainly un-private [this `PostSerivceRESTMock` type](https://github.com/wordpress-mobile/WordPress-iOS/blob/0c8b0319329f90bf838c5fa101c1cb6e6a901773/WordPress/WordPressTest/PostRepositoryTests.swift#L332) and use it here. But now the responsibility of creating fake remote posts (or error) falls to each tests.

BTW, you can see the difference in how fake remotes posts are created comparing tests in this PR to [the other PR][OPR].

However, I'm not saying one is better than the other. All I said above it pretty much in context of this PR, where the test cases just needs an API to return either some posts or an error.

[OPR]: https://github.com/wordpress-mobile/WordPress-iOS/pull/21999/files#diff-7288f0e3763e9cbfb4e8305fbfc49ff6f3bdbe975355053a68f88f3636d8c0aa.

File: WordPress/WordPressTest/PostRepositoryPostsListTests.swift (hunk 3 above)
Comment (crazytonyli) [reply WordPress/WordPressTest/PostRepositoryPostsListTests.swift]: I have removed it in https://github.com/wordpress-mobile/WordPress-iOS/pull/21997/commits/cfe7a675ba31ce724a3b180bb4420df87537730f. It's duplicated with other tests.

**Comment by crazytonyli** [comment]
I managed to fix [the three failed UI tests](https://buildkite.com/automattic/wordpress-ios/builds/18269#annotation-test-summary-018bb175-461c-4be2-ae7c-9bfa78dbbeca-have-failed). But I have found some interesting things in the API stubs setup. I thought I'll share them here, just in case they are useful to @wordpress-mobile/mobile-ui-testing-squad.

> 3 tests have failed (3 distinct assertion failure(s) in total)
> - testCreateBlankPage() in PageTests
> - testBasicPostPublishWithCategoryAndTag() in EditorGutenbergTests (3 times)
> - testTextPostPublish() in EditorGutenbergTests (3 times)

## App changes

To start, I'll describe what's changed in the app in terms of API requests it sends.

Before this change, when the dashboard loads page cards, it sends at least two API requests to fetch all pages data from the site. This is a bug (see #21812), because dashboard doesn't need all the pages, it just needs three. With this PR's change, the dashboard sends one API request to get three page instances.

There is also one small change in the actual API request content. Before this change, the very first fetching pages API request's URL looks like something like `sites/181851495/posts?number=3&type=page`. With this PR's change, the request URL adds a `offset=0` query, which doesn't really matter to the actual backend server. But I think it may have changed how WireMock picks an API stub form various json files.

## The failing UI tests

### testCreateBlankPage

The cause of this failure is date fields in `sites_181851495_pages_after.json` are in a wrong format. The app expects a timezone offset in it, which the API stub doesn't have. This issue is fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21997/commits/629ed5d33d5af1524985a2960284e7397d84ced1.

This obviously raises a question: why this test doesn't fail on the trunk branch. The only reason I can think of is the trunk branch uses other API stub instead the one declared in `sites_181851495_pages_after.json`. Based on the secnario and state transition declared in the file, I think `testCreateBlankPage` is supposed to use `sites_181851495_pages_after.json` but it didn't? So maybe the trunk branch should not even pass?

https://github.com/wordpress-mobile/WordPress-iOS/blob/1511635f090a05a13b9983e0aa598332f62402f0/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/pages/sites_181851495_pages_after.json#L2-L4

### testBasicPostPublishWithCategoryAndTag

This test passed on my mac. I didn't dive deeper into why it failed on CI.

### testTextPostPublish

The cause of this failure is WireMock reports a 'Scenario not match' error on `POST posts/new` API request, which causes the app failing to publish the post.

Here is my understand of how the API stubs setup _should_ work:

First, `posts_draft,pending.json` is picked up when the app launches into dashboard, which transits the state to `initial_posts_list`.

https://github.com/wordpress-mobile/WordPress-iOS/blob/1511635f090a05a13b9983e0aa598332f62402f0/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/posts/posts_draft%2Cpending.json#L2-L4

Later, after UI test taps post editor's Publish button, the `post_new.json` file gets picked up and returns a result that indicates the post has been published.

https://github.com/wordpress-mobile/WordPress-iOS/blob/1511635f090a05a13b9983e0aa598332f62402f0/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/posts/posts_new.json#L2-L4

However, during the test run, the first step `posts_draft,pending.json` didn't get picked up. The root cause is the dashboard didn't show a post card, based on this API stub response.

https://github.com/wordpress-mobile/WordPress-iOS/blob/1511635f090a05a13b9983e0aa598332f62402f0/API-Mocks/WordPressMocks/src/main/assets/mocks/mappings/wpcom/dashboard/106707880%20-%20Tri-County%20Real%20Estate/sites_106707880_dashboard.json#L23-L27

This issue is fixed by adding a draft post to the API stub. See https://github.com/wordpress-mobile/WordPress-iOS/pull/21997/commits/44b9ce08b89728029cc32380c9f713f4a70b2ea4.

I verified that, on the trunk branch, the dashboard also doesn't show post card and `posts_draft,pending.json` doesn't get picked up by WireMock. I tried to understand which API stub triggered the state transition to `initial_posts_list` and allows `post_new.json` to pass on trunk branch, but I'm not familiar how WireMock works and how the declared API stubs work together for each UI test. I'm interested to learn though, if there is any documentation about it.

----

Both fixes I made above are not related to the actually HTTP API request changes that's caused by this PR. If my understanding is correct, they are due to incorrect API stubs set up. Maybe this is something @wordpress-mobile/mobile-ui-testing-squad you'd be interested in diving deeper?

**Comment by jostnes** [comment]
Thanks for sharing your findings about the failing tests on the branch @crazytonyli! And for the fixes to make them 🟢 again! 

It's strange that it's failing on the branch but not on trunk when the test failures are not related to the changes in the PR. I'm not an expert but I've been using their [official doc for reference](https://wiremock.org/2.x/docs/stubbing/). I like to point out that the project is still using v2.x of WireMock (latest is v3.x) - I'm not sure how big of an effort it's going to be to update it to v3.x and if the oddities are because we're using an older version. 

I can probably look at this sometime next sprint and let you know if I find anything else that you didn't already.

**Pull request merged**